File: GKPageScrollViewObjc/GKPageScrollViewObjc/Demo/Nest/GKNest2ViewController.m
//
//  GKNest2ViewController.m
//  GKPageScrollViewObjc
//
//  Created by gaokun on 2019/10/21.
//  Copyright © 2019 gaokun. All rights reserved.
//

#import "GKNest2ViewController.h"
#import "GKNestScrollView.h"
#import "GKNest2View.h"
#import <JXCategoryView/JXCategoryView.h>

@interface GKNest2ViewController ()<JXCategoryViewDelegate, UIScrollViewDelegate, GKNestScrollViewGestureDelegate>

@property (nonatomic, strong) JXCategoryTitleView *categoryView;

@property (nonatomic, strong) GKNestScrollView    *contentScrollView;

@property (nonatomic, weak) GKNest2View           *nestView;

@end

@implementation GKNest2ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.gk_backStyle = GKNavigationBarBackStyleBlack;
    self.gk_navTitleView = self.categoryView;
    
    [self.view addSubview:self.contentScrollView];
    
    [self.contentScrollView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.bottom.equalTo(self.view);
        make.top.equalTo(self.gk_navigationBar.mas_bottom);
    }];
    
    [self categoryView:self.categoryView didSelectedItemAtIndex:0];
}

#pragma mark - JXCateogryViewDelegate
- (void)categoryView:(JXCategoryBaseView *)categoryView didSelectedItemAtIndex:(NSInteger)index {
    self.nestView = (GKNest2View *)self.contentScrollView.subviews[index];
}

#pragma mark - GKNestScrollViewGestureDelegate
- (BOOL)nestScrollView:(GKNestScrollView *)scrollView gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    if ([self panBackWithScrollView:scrollView gestureRecognizer:gestureRecognizer]) {
        return NO;
    }
    
    UIScrollView *listScrollView = self.nestView.pageScrollView.listContainerView.collectionView;
    
    if (listScrollView.isTracking || listScrollView.isDragging) {
        if ([gestureRecognizer isMemberOfClass:NSClassFromString(@"UIScrollViewPanGestureRecognizer")]) {
            CGFloat velocityX = [(UIPanGestureRecognizer *)gestureRecognizer velocityInView:gestureRecognizer.view].x;
            // x大于0就是右滑，如果列表容器没有滑动到最右边，禁止滑动
            if (velocityX > 0) {
                if (listScrollView.contentOffset.x != 0) {
                    return NO;
                }
            }else if (velocityX < 0) { // x小于0是往左滑
                if (listScrollView.contentOffset.x + listScrollView.bounds.size.width != listScrollView.contentSize.width) {
                    return NO;
                }
            }
        }
    }
    
    return YES;
}

- (BOOL)nestScrollView:(GKNestScrollView *)scrollView gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    if ([self panBackWithScrollView:scrollView gestureRecognizer:gestureRecognizer]) return YES;
    
    return NO;
}

// 检查contentScrollView是否滑动到边缘
- (BOOL)checkIsContentScrollEdge {
    UIScrollView *listCollectionView = self.nestView.pageScrollView.listContainerView.collectionView;
    
    if (listCollectionView.contentOffset.x == 0 || (listCollectionView.contentOffset.x + listCollectionView.frame.size.width) == listCollectionView.contentSize.width) {
        return YES;
    }
    return NO;
}

- (BOOL)panBackWithScrollView:(UIScrollView *)scrollView gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer {
    if (gestureRecognizer == scrollView.panGestureRecognizer) {
        CGPoint point = [scrollView.panGestureRecognizer translationInView:scrollView];
        UIGestureRecognizerState state = gestureRecognizer.state;
        
        // 设置手势滑动的位置距屏幕左边的区域
        CGFloat locationDistance = [UIScreen mainScreen].bounds.size.width;
        
        if (state == UIGestureRecognizerStateBegan || state == UIGestureRecognizerStatePossible) {
            CGPoint location = [gestureRecognizer locationInView:scrollView];
            if (point.x > 0 && location.x < locationDistance && scrollView.contentOffset.x <= 0) {
                return YES;
            }
        }
    }
    return NO;
}

#pragma mark - 懒加载
- (JXCategoryTitleView *)categoryView {
    if (!_categoryView) {
        _categoryView = [[JXCategoryTitleView alloc] initWithFrame:CGRectMake(0, 0, kScreenW, 44.0f)];
        _categoryView.titles = @[@"精选", @"时尚", @"电器", @"超市", @"生活", @"运动", @"饰品", @"数码", @"家装", @"手机"];
        _categoryView.titleFont = [UIFont systemFontOfSize:15.0f];
        _categoryView.titleSelectedFont = [UIFont systemFontOfSize:16.0f];
        _categoryView.titleColor = [UIColor blackColor];
        _categoryView.titleSelectedColor = [UIColor blackColor];
        _categoryView.delegate = self;
        _categoryView.contentEdgeInsetLeft = 0;
        _categoryView.contentEdgeInsetRight = 0;
        
        JXCategoryIndicatorLineView *lineView = [JXCategoryIndicatorLineView new];
        lineView.lineStyle = JXCategoryIndicatorLineStyle_Lengthen;
        _categoryView.indicators = @[lineView];
        
        _categoryView.contentScrollView = self.contentScrollView;
    }
    return _categoryView;
}

- (GKNestScrollView *)contentScrollView {
    if (!_contentScrollView) {
        _contentScrollView = [GKNestScrollView new];
        _contentScrollView.showsVerticalScrollIndicator = NO;
        _contentScrollView.showsHorizontalScrollIndicator = NO;
        _contentScrollView.delegate = self;
        _contentScrollView.gestureDelegate = self;
        _contentScrollView.pagingEnabled = YES;

        CGFloat width = kScreenW;
        CGFloat height = kScreenH - GK_STATUSBAR_HEIGHT - 50.0f;

        [self.categoryView.titles enumerateObjectsUsingBlock:^(NSString * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            GKNest2View *nestView = [GKNest2View new];
            nestView.frame = CGRectMake(idx * width, 0, width, height);
            nestView.mainScrollView = self->_contentScrollView;
            [self->_contentScrollView addSubview:nestView];
        }];

        _contentScrollView.contentSize = CGSizeMake(self.categoryView.titles.count * width, 0);
    }
    return _contentScrollView;
}

@end

File: GKPageScrollViewSwift/GKPageScrollViewSwift/Demo/Nest/GKNest2ViewController.swift
//
//  GKNest2ViewController.swift
//  GKPageScrollViewSwift
//
//  Created by gaokun on 2019/10/22.
//  Copyright © 2019 gaokun. All rights reserved.
//

import UIKit
import JXSegmentedView

class GKNest2ViewController: GKDemoBaseViewController {

    var titleDataSource = JXSegmentedTitleDataSource()
    
    var currentNestView = GKNest2View()
    
    lazy var segmentedView: JXSegmentedView = {
        titleDataSource.titleNormalColor = UIColor.black
        titleDataSource.titleSelectedColor = UIColor.black
        titleDataSource.titleNormalFont = UIFont.systemFont(ofSize: 15.0)
        titleDataSource.titleSelectedFont = UIFont.systemFont(ofSize: 16.0)
        titleDataSource.titles = ["精选", "时尚", "电器", "超市", "生活", "运动", "饰品", "数码", "家装", "手机"]
        
        var segmentedView = JXSegmentedView(frame: CGRect(x: 0, y: 0, width: kScreenW, height: 50.0))
        segmentedView.delegate = self
        segmentedView.dataSource = titleDataSource
        segmentedView.contentEdgeInsetLeft = 0;
        segmentedView.contentEdgeInsetRight = 0;
        
        let lineView = JXSegmentedIndicatorLineView()
        lineView.lineStyle = .normal
        lineView.indicatorHeight = ADAPTATIONRATIO * 4.0
        lineView.verticalOffset = ADAPTATIONRATIO * 2.0
        segmentedView.indicators = [lineView]
        
        segmentedView.contentScrollView = self.contentScrollView
        
        return segmentedView
    }()
    
    lazy var contentScrollView: GKNestScrollView = {
        let scrollView = GKNestScrollView()
        scrollView.showsVerticalScrollIndicator = false
        scrollView.showsHorizontalScrollIndicator = false
        scrollView.delegate = self
        scrollView.isPagingEnabled = true
        scrollView.gestureDelegate = self
        
        let width = kScreenW
        let height = kScreenH - kStatusBar_Height - 50.0
        
        for (i, value) in self.titleDataSource.titles.enumerated() {
            let nestView = GKNest2View()
            nestView.mainScrollView = scrollView
            nestView.frame = CGRect(x: CGFloat(i) * width, y: 0, width: width, height: height)
            scrollView.addSubview(nestView)
        }
        scrollView.contentSize = CGSize(width: CGFloat(self.titleDataSource.titles.count) * width, height: 0)
        return scrollView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.gk_backStyle = .black
        
        self.view.addSubview(self.segmentedView)
        self.view.addSubview(self.contentScrollView)
        
        self.segmentedView.frame = CGRect(x: 0, y: 0, width: kScreenW, height: 44.0)
        self.gk_navTitleView = self.segmentedView
        
        self.contentScrollView.snp.makeConstraints { (make) in
            make.left.right.bottom.equalTo(self.view)
            make.top.equalTo(self.gk_navigationBar.snp_bottom)
        }
        
        self .segmentedView(self.segmentedView, didSelectedItemAt: 0)
    }
}

extension GKNest2ViewController: JXSegmentedViewDelegate {
    func segmentedView(_ segmentedView: JXSegmentedView, didSelectedItemAt index: Int) {
        self.currentNestView = self.contentScrollView.subviews[index] as! GKNest2View
    }
}

extension GKNest2ViewController: UIScrollViewDelegate {}
extension GKNest2ViewController: GKNestScrollViewDelegate {
    func nestScrollViewGestureRecognizerShouldBegin(_ scrollView: GKNestScrollView, gestureRecognizer: UIGestureRecognizer) -> Bool {
        if self.panBack(scrollView: scrollView, gestureRecognizer: gestureRecognizer) {
            return false
        }
        
        if let listScrollView = self.currentNestView.pageScrollView.listContainerView.collectionView {
            if listScrollView.isTracking || listScrollView.isDragging {
                if gestureRecognizer.isMember(of: NSClassFromString("UIScrollViewPanGestureRecognizer")!) {
                    let panGestureRecognizer = gestureRecognizer as! UIPanGestureRecognizer
                    
                    let velocityX = panGestureRecognizer.velocity(in: gestureRecognizer.view).x
                    
                    if velocityX > 0 {
                        if listScrollView.contentOffset.x != 0 {
                            return false
                        }
                    }else if velocityX < 0 {
                        if listScrollView.contentOffset.x + listScrollView.bounds.size.width != listScrollView.contentSize.width {
                            return false
                        }
                    }
                }
            }
        }
        
        return true
    }
    
    func nestScrollGestureRecognizer(_ scrollView: GKNestScrollView, gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        return false
    }
    
    func panBack(scrollView: UIScrollView, gestureRecognizer: UIGestureRecognizer) -> Bool{
        if gestureRecognizer == scrollView.panGestureRecognizer {
            let point = scrollView.panGestureRecognizer.translation(in: scrollView)
            let state = gestureRecognizer.state
            
            let locationDistance = UIScreen.main.bounds.size.width
            
            if state == .began || state == .possible {
                let location = gestureRecognizer.location(in: scrollView)
                if point.x > 0 && location.x < locationDistance && scrollView.contentOffset.x <= 0 {
                    return true
                }
            }
        }
        return false
    }
}
